Modules/Loadable/CropVolume/Logic/vtkSlicerCropVolumeLogic.cxx

```diff
@@ -114,563 +121,649 @@ void vtkSlicerCropVolumeLogic::PrintSelf(ostream& os, vtkIndent indent)
     }
-
-  std::ostringstream outSS;
-  double outputSpacing[3], spacingScaleConst = pnode->GetSpacingScalingConst();
-  outSS << inputVolume->GetName() << "-subvolume-scale_" << spacingScaleConst;
-
-  if(dtvnode)
+  if (vtkMRMLDiffusionTensorVolumeNode::SafeDownCast(inputVolume))
     {
-    std::cerr << "CropVolume: ERROR: Diffusion tensor volumes are not supported by this module!" << std::endl;
+    vtkErrorMacro("CropVolume: Diffusion tensor volumes are not supported by this module");
     return -2;
     }
-  // need to create clones and display nodes here, since
-  // VolumesLogic::CloneVolume() handles only ScalarVolumeNode's
-  else if(dwvnode)
-    {
-    vtkNew<vtkMRMLDiffusionWeightedVolumeNode> outputDWVNode;
-    outputDWVNode->CopyWithScene(dwvnode);
-    vtkNew<vtkMRMLDiffusionWeightedVolumeDisplayNode> dwiDisplayNode;
-    dwiDisplayNode->CopyWithScene(dwvnode->GetDisplayNode());
-    scene->AddNode(dwiDisplayNode.GetPointer());
-
-    vtkNew<vtkImageData> outputImageData;
-    outputImageData->DeepCopy(dwvnode->GetImageData());
-    outputDWVNode->SetAndObserveImageData(outputImageData.GetPointer());
-
-    outputDWVNode->SetAndObserveDisplayNodeID(dwiDisplayNode->GetID());
-    outputDWVNode->SetAndObserveStorageNodeID(NULL);
-    scene->AddNode(outputDWVNode.GetPointer());
 
-    outputVolume = outputDWVNode.GetPointer();
-    }
-  else if(vvnode)
-    {
-    vtkNew<vtkMRMLVectorVolumeNode> outputVVNode;
-    outputVVNode->CopyWithScene(dwvnode);
-    vtkNew<vtkMRMLVectorVolumeDisplayNode> vvDisplayNode;
-    vvDisplayNode->CopyWithScene(vvnode->GetDisplayNode());
-    scene->AddNode(vvDisplayNode.GetPointer());
-
-    vtkNew<vtkImageData> outputImageData;
-    outputImageData->DeepCopy(vvnode->GetImageData());
-    outputVVNode->SetAndObserveImageData(outputImageData.GetPointer());
-
-    outputVVNode->SetAndObserveDisplayNodeID(vvDisplayNode->GetID());
-    outputVVNode->SetAndObserveStorageNodeID(NULL);
-    scene->AddNode(outputVVNode.GetPointer());
-
-    outputVolume = outputVVNode.GetPointer();
-    }
-  else if(svnode)
+  // Check/create output volume
+  vtkMRMLVolumeNode *outputVolume =
+    vtkMRMLVolumeNode::SafeDownCast(scene->GetNodeByID(pnode->GetOutputVolumeNodeID()));
+  if (outputVolume)
     {
-    outputVolume = this->Internal->VolumesLogic->CloneVolume(this->GetMRMLScene(), inputVolume, outSS.str().c_str());
+    // Output volume is provided, use that (if compatible)
+    if (outputVolume->GetClassName() != inputVolume->GetClassName())
+      {
+      vtkErrorMacro("CropVolume: output volume (" << outputVolume->GetClassName() <<
+        ") is not compatible with input volume (" << inputVolume->GetClassName() << ")");
+      return -1;
+      }
     }
   else
     {
-    std::cerr << "Input volume not recognized!" << std::endl;
-    return -1;
+    // Create compatible output volume
+    if (!this->Internal->VolumesLogic)
+      {
+      vtkErrorMacro("CropVolume: invalid Volumes logic");
+      return -2;
```

Comment:
I just kept the existing error codes to preserve backward compatibility but you are right, at least we should document the return values.